File: chapter10/src/main/java/com/example/chapter10/part1/BitmapFactoryDecodeResourceView.java
package com.example.chapter10.part1;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import com.example.chapter10.R;

/**
 * @author wangzhichao
 * @date 2019/10/20
 */
public class BitmapFactoryDecodeResourceView extends View {
    private Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private Bitmap bitmap;
    public BitmapFactoryDecodeResourceView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.scenery);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        canvas.drawBitmap(bitmap,null, new Rect(0,0,getWidth(),getHeight()),paint);
    }
}

File: chapter10/src/main/java/com/example/chapter10/part2/BitmapFactoryDecodeResourceView.kt
package com.example.chapter10.part2

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import com.example.chapter10.R

/**
 * @author wangzhichao
 * @date 2019/10/20
 */
class BitmapFactoryDecodeResourceView(context: Context, attrs: AttributeSet?) :
    View(context, attrs) {
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val bitmap: Bitmap
    init {
        bitmap = BitmapFactory.decodeResource(resources, R.drawable.scenery)
    }
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawBitmap(bitmap, null, Rect(0, 0, width, height), paint)
    }
}